### ISSUE TITLE
Error message step: <Tween#->: started with no Tweeners missing essential information

### ISSUE THREAD

**Issue by 2fd5** [godotengine/godot#79055]
### Godot version

v4.1.rc2.mono.official [46424488e]

### System information

Linux

### Issue description

Getting error message at runtime.
The error message does not show what is the origin of this error.

```
E 0:00:36:0394   step: <Tween#-9223371483558965298>: started with no Tweeners.
  <C++ Error>    Method/function failed. Returning: false
  <C++ Source>   scene/animation/tween.cpp:311 @ step()
```

Expected behaviour would be to have error message show which tween, meaning, which .gd script line that is attempting to start a tween that is not created. So that it is easier to fix the issue. Now I would have to go through all tweens in the game.

### Steps to reproduce

Not sure if that is required to fix the issue by providing more information in error message, to allow for tracking it to specific origin of the call.

### Minimal reproduction project

none

**Comment by KoBeWi** [comment]
We currently don't have any means to track where an object was created. However if you bind a Tween to a node, the node will be displayed in the error message.

**Comment by 2fd5** [comment]
Does it matter where the object was created?. All I was expecting is to display the GDscript file name and line number which caused the error message to appear. In other words
If the error is caused when `tween.tween_property` is called then it would be helpful to know the filename and line number of that `tween.tween_property`. It is enough to find the cause of the issue. 
I do not need to know where `tween` was created. 

It a matter of what is calling objects `tween_property` with disregard of what created the object.

Am I missing some crucial implementation detail of how tweens are created and called ?
Thanks

**Comment by AThousandShips** [comment]
It is started outside of the script, at the end of the current frame I think, so it's not started at any one line of the script

**Comment by KoBeWi** [comment]
Well, you can manually call `play()` to start a Tween, in which case the error could be tracked (not tested). The Tweens that start automatically do so from C++ code, you can see the C++ line in the error.

**Comment by 2fd5** [comment]
Ok so in my case the error message was caused by having
```
@onready var highligt_tween: Tween = get_tree().create_tween()
```

At the scope of GDscript class that was used by the Node2D scene. Is this supported use case to have tweens created in `_ready()` and then reused ?

**Comment by AThousandShips** [comment]
What happens if you remove `@onready`

**Comment by 2fd5** [comment]
> Yes, the tween starts automatically, so just creating it and not initializing it is incorrect
Is this consistent with other objects in Godot ?
Does it need any additional documentation ?

Not sure that
```
A Tween can be created by using either [SceneTree.create_tween](https://docs.godotengine.org/en/stable/classes/class_scenetree.html#class-scenetree-method-create-tween) or [Node.create_tween](https://docs.godotengine.org/en/stable/classes/class_node.html#class-node-method-create-tween). Tweens created manually (i.e. by using Tween.new()) are invalid and can't be used for tweening values.
```

And 
```
Note: All Tweens will automatically start by default. To prevent a Tween from autostarting, you can call [stop](https://docs.godotengine.org/en/stable/classes/class_tween.html#class-tween-method-stop) immediately after it is created.
```
Actually indicates that creating and not using is incorrect. From what I can understand it says, use `SceneTree.create_tween` to create and call `stop` to prevent it from starting automatically.

**Comment by AThousandShips** [comment]
Yes, it's correct to stop it, but your usage doesn't do that, that's what I mean is incorrect

**Comment by 2fd5** [comment]
> Yes, it's correct to stop it, but your usage doesn't do that, that's what I mean is incorrect, hence the error
> 
> The two correct usages are:
> 
> * Create and use immediately
> * Create, stop, use later and call `play`

Do you think documentation could be improved ?

I would like to start contributing to Godot docs. I will try to get the setup in place to do so. I might use this opportunity to start with tweaking tween class doc if you guys think there's a need for that.

Or did I just not understood the docs in the first place ?

**Comment by AThousandShips** [comment]
How would you suggest an improvement? What beyond "they start automatically" would be needed here in your opinion?

**Comment by 2fd5** [comment]
> How would you suggest an improvement? What beyond "they start automatically" would be needed here in your opinion?

For example just to add something like:
If you are not planning to use the tween immediately withing the scope call `.stop()` when initialising.

```
@onready tween: Tween = get_tree().create_tween().stop()
```

**Comment by 2fd5** [comment]
> Also you can't use that code `stop` doesn't return a `Tween`, so your suggestion won't work

Yes just noticed that it would rather be.
```
@onready tween: Tween

func _ready():
    tween = get_tree().create_tween()
    tween.stop()

func fade_out():
    if not tween.is_running():
        tween.tween_property(self, "modulate", Color(1, 1, 1, 1.0), 0.4)
```

Well it was not obvious to me. Not sure if that is just me or more users that might encounter that. 

I am not running this project so it is not my call.

It's the matter of who is being targeted as the potential Godot user and what are the expectations from them.

GDevelop tries to be easy for newcomers. I would imagine they would add that to the docs or change the behaviour. I might be wrong.

I do understand that the need to keep the docs short and to the point.

I expect some guidelines and help here on what to document and what not. Thanks for your time.

**Comment by KoBeWi** [comment]
> The two correct usages are:
- Create, stop, use later and call play

Actually it's not exactly correct. Tweens are heavily encouraged to be setup immediately on create and never reused (#74258). The code might be valid, but doesn't make much practical sense.

Although we recently had a discussion to maybe allow re-using Tweens. While the is no real performance gain, it can *sometimes* save a bit of boilerplate code. The `create_tween()` function would get an `autostart` argument to make it easier. But it was only discussed, no decision was made so far (and there is no proposal).

**Comment by AThousandShips** [comment]
Agreed, more that those are the two methods that aren't incorrect i.e. throws errors

**Comment by 2fd5** [comment]
Not sure what is the suggested design patter / propositions for user that wants to tween same property from different places and do not want to interrupt already running tweens.
I understand that it is not encouraged be cause implementation details but what is being offered for these use-cases ?
Other then suggested  https://github.com/godotengine/godot/pull/74258 `autostart` argument ?

**Comment by AThousandShips** [comment]
Getting somewhat of topic here though

The solution here is that this issue is due to misuse of the `Tween`, which at least in the sense that it automatically starts is documented, if anything the documentation should IMO be changed to emphasize the appropriate usage

**Comment by 2fd5** [comment]
> Not sure what is the suggested design patter / propositions for user that wants to tween same property from different places and do not want to interrupt already running tweens. I understand that it is not encouraged be cause implementation details but what is being offered for these use-cases ? Other then suggested #74258 `autostart` argument ?

Yes, this might be off topic for this  issue at hand. But IMHO  not for the tweens, the documentation and the relations to the community.

**Comment by AThousandShips** [comment]
Therefore better handled in an appropriate proposal

But the discussion I said was on topic, so I'm not sure what you are responding to there

**Comment by 2fd5** [comment]
> Getting somewhat of topic here though

I was responding to above. Sorry it that was unclear. Thank you I will comment on the suggested proposal in regards to alternative patter for not reusing the tweens.

In regards to this issue. Can I close it with comment that documentation could possibly be improved, or shall I leave to some triage meeting ?

**Comment by AThousandShips** [comment]
Leave it open to discuss the documentation, we don't have triage meetings :)

**Comment by Fox-Alpha** [comment]
> > How would you suggest an improvement? What beyond "they start automatically" would be needed here in your opinion?
> 
> For example just to add something like: If you are not planning to use the tween immediately withing the scope call `.stop()` when initialising.
> 
> ```
> @onready tween: Tween = get_tree().create_tween().stop()
> ```

When i used the suggested line, i become this Error

`Cannot get return value of call to "stop()" because it returns "void".`

**Comment by 2fd5** [comment]
> 

yes calling stop when assigning tween to a variable won't work as `stop()` does not return anything

it looks like it is better to just define variable
`var tween: Tween` and then create tween when needed. or `tween = get_tree().create_tween()` and in next line stop the tween `tween.stop()`

I think [documentation](https://docs.godotengine.org/en/stable/classes/class_tween.html) already cover the use case that I was struggling with

>  You should avoid using more than one Tween per object's property. If two or more tweens animate one property at the same time, the last one created will take priority and assign the final value. If you want to interrupt and restart an animation, consider assigning the Tween to a variable:
```
var tween
func animate():
    if tween:
        tween.kill() # Abort the previous animation.
    tween = create_tween()
```

**Comment by alex-dearden** [comment]
I would second the idea of adding something to the docs along the proper usage as suggested here:

```
onready tween: Tween

func _ready():
    tween = get_tree().create_tween()
    tween.stop()
```

I ended up on this page because I was getting the same error, even when creating the tween in the method that used it. 

While searching the web for this error, I saw a lot of other people asking the same thing and baffled by why their tweens weren't working with no good answers. As mentioned by the OP, GDScript is geared to be easy for newcomers, most of whom will not read GitHub pages and this seems to be the only place were I could find a good answer.

**Comment by KoBeWi** [comment]
Tweens are not designed for reusing. The documentation warns about that and none of the examples point to creating a Tween with intention of later usage. The example you suggested actually existed until #74258

**Comment by flotos** [comment]
Hi, from what I see this error should happen when the tweener isn't used immediately. However in my case I use it on creation :

```
func _ready():
	start_combat()

func entity_scale(entity_sprite: Node2D, f_scale: float):
	entity_sprite.scale = Vector2(f_scale, f_scale)

func animate_spawn():
	var tween_test = get_tree().create_tween().set_parallel(true).set_ease(Tween.EASE_OUT)
	tween_test.tween_method(entity_scale, enemy, 3.5, 1)

func start_combat():
	animate_spawn()
```

and with that I get the same error (which was verified to come from this snippet). However from my understanding the use of the tween is instant and not reused.

**Comment by KoBeWi** [comment]
Your usage of `tween_method()` is wrong and it causes invalid Tweener. You should be getting an error about it, although.

**Comment by pixl-wyrm** [comment]
Maybe I'm alone in this, but I still think this error message is rather vague, partly because "Tween" and "Tweener" are very similar words. I've referenced the Tween documentation quite a lot, but my first instinct for a long time was to think "What do you mean the Tween has no Tweeners? You ARE a Tweener!" and I only recently realized that they are not the same thing. Granted, I'm not the most technically-minded, but I think this particular error message is a pain point for beginners that could be easily cleared up, even if it can't include the specific line that caused the error.

I think the shortest, simplest clarification would be "started with no _subsequent_ Tweeners", but you could also phrase it as "started with no Tweeners and instantly finished", which explains why you can't just leave a Tween running with nothing to do. You could also append this line from the docs: "Only create a Tween when you want to start animating." but perhaps that would bloat the message too much.
